### PR TITLE
sd-radv: several cleanups for router lifetime

### PR DESCRIPTION
**Pull request by yuwata**
Split out of #21112.

### CODE REVIEW

File: src/libsystemd-network/sd-radv.c [hunk 1]
```diff
@@ -167,13 +168,21 @@ static int radv_send(sd_radv *ra, const struct in6_addr *dst, uint32_t router_li
         if (r < 0)
                 return r;
 
+        /* a value of UINT16_MAX represents infinity, 0x0 means this host is not a router */
+        if (lifetime_usec == USEC_INFINITY)
+                lifetime_sec = UINT16_MAX;
+        else if (lifetime_usec > (UINT16_MAX - 1) * USEC_PER_SEC)
```
Comment (poettering) [inline src/libsystemd-network/sd-radv.c]: doesn't matter, but >= would be less surprising, i.e. it has more of a "this is saturation" feeling to me

File: src/libsystemd-network/sd-radv.c [hunk 2]
```diff
@@ -159,31 +165,23 @@ static int radv_send(sd_radv *ra, const struct in6_addr *dst, usec_t lifetime_us
...
-                lifetime_sec = DIV_ROUND_UP(lifetime_usec, USEC_PER_SEC);
-
         if (dst && in6_addr_is_set(dst))
                 dst_addr.sin6_addr = *dst;
 
         adv.nd_ra_type = ND_ROUTER_ADVERT;
         adv.nd_ra_curhoplimit = ra->hop_limit;
         adv.nd_ra_flags_reserved = ra->flags;
-        adv.nd_ra_router_lifetime = htobe16(lifetime_sec);
+        adv.nd_ra_router_lifetime = htobe16(DIV_ROUND_UP(lifetime_usec, USEC_PER_SEC));
```
Comment (poettering) [inline src/libsystemd-network/sd-radv.c:185]: maybe add:

```c
assert_cc(DIV_ROUND_UP(SD_RADV_MAX_ROUTER_LIFETIME_USEC, USEC_PER_SEC) < UINT16_MAX);
```

Just to document that this cannot overflow

File: src/libsystemd-network/radv-internal.h [hunk 3]
```diff
@@ -12,7 +12,8 @@
 #include "sparse-endian.h"
 #include "time-util.h"
 
-assert_cc(SD_RADV_DEFAULT_MIN_TIMEOUT_USEC <= SD_RADV_DEFAULT_MAX_TIMEOUT_USEC);
+#define SD_RADV_DEFAULT_MAX_TIMEOUT_USEC             (600 * USEC_PER_SEC)
+#define SD_RADV_DEFAULT_MIN_TIMEOUT_USEC             (200 * USEC_PER_SEC)
```
Comment (poettering) [inline src/libsystemd-network/radv-internal.h]: the SD_ prefix is for exported stuff, should probably dropped for all stuff defined in radv-internal.h, given that that's not exported

File: src/network/networkd-radv.c [hunk 4]
```diff
@@ -1256,6 +1256,49 @@ int config_parse_router_prefix_delegation(
...
+
+        r = parse_sec(rvalue, &usec);
+        if (r < 0) {
+                log_syntax(unit, LOG_WARNING, filename, line, r,
+                           "Failed to parse router lifetime, ignoring assignment: %s", rvalue);
+                return 0;
+        }
+        if (usec > 0 &&
+            (usec < SD_RADV_MIN_ROUTER_LIFETIME_USEC || usec > SD_RADV_MAX_ROUTER_LIFETIME_USEC)) {
+                log_syntax(unit, LOG_WARNING, filename, line, r,
```
Comment (poettering) [inline src/network/networkd-radv.c]: r is not set here

File: src/network/networkd-radv.c [hunk 5]
```diff
@@ -1256,6 +1256,49 @@ int config_parse_router_prefix_delegation(
...
+                           "Invalid router lifetime, ignoring assignment: %s", rvalue);
+                return 0;
+        }
```
Comment (poettering) [inline src/network/networkd-radv.c:1306]: wouldn't it make more sense to just clamp the vaues to the right range? i.e. clamp, log and continue? i mean the intention of the user is kinda clear: "i want a super low value" or "i want a super high value", and this isn't a super critical value, so i'd handle this gracefully.

File: src/libsystemd-network/sd-radv.c (hunk 1 above)
Comment (yuwata) [reply src/libsystemd-network/sd-radv.c]: Yeah, but this condition is copied from networkd-radv.c, and is removed by later commit. I'd like to keep this as is.

**Pull request merged**